firmware/gameport-adapter/ThrustmasterAttackThrottle.h

```diff
@@ -0,0 +1,370 @@
+  /// an enable line and another trigger pulse. I believe that is useful with more
+  /// than 1 digital joystick chained together. I only have one to test so I ignored it.
+  ///
+  /// The Linux kernel code is pretty much useless. I found it easier to reverse engineer
+  /// myself. There is minimal documentation, single-letter variable names and just
+  /// a complicated structure. The protocol is so simple it was easier to capture
+  /// and process myself.
+  ///
+  /// @param packet is a buffer that is filled with the packet data
+  /// @returns true if the data read ok, false on an error
+  bool readPacket(uint8_t *packet) {
+
+    // I'm not sure if this optimization actually does anything on AVR.
+    // My thought is locality of reference may help to get the core loop
+    // to be more performant. This is memcpy'ed to packet at the end when timing
+    // doesn't matter.
+    uint8_t localPacket[packetSize];
+
+    // Halt interrupts only during the timing-sensitve part of the code
+    //
+    // This code is *very* timing-sensitive. There was a time where it was
+    // working reliably until I changed an unrelated part of the code. Not
+    // sure if that impacted inlining or code layout or something but it was
+    // enough to break the timing.
+    // I think this version is reliable. It has been in practice
+    // but I wouldn't rule out something else odd like this happening again.
+    const InterruptStopper interruptStopper;
+    {
+      trigger();
+
+      // the first clock has a 200us timeout, every following clock
+      // has a much shorter timeout. I've seen the gap between the
+      // trigger and the first clock be about 100us. So 200us should
+      // be plenty of margin. Having a long timeout on the first bit
+      // makes this simpler than having extra wait calls to account
+      // for this gap.
+      int timeout = 200;
+
+      // each byte is shifted into this variable
+      uint8_t shiftReg;
+
+      // current location in filling in the packet
+      uint8_t *buffer = localPacket;
+
+      for (uint8_t byte = 0; byte < packetSize; ++byte) {
+        shiftReg = 0;
+
+        // check for the start bit, it should always be 1
+        if (!m_clock.wait(Edge::falling, timeout)) {
+          log("timeout clk");
+          return false;
+        }
+
+        if (!m_data.get()) {
+          // bad start bit
+          return false;
+        }
+
+        // after the first bit in the packet, adjust the timeout.
+        // A clock pulse is 12us. so this should be enough although
+        // I don't know how precise the timeout is.
+        timeout = 24;
+
+        // the data portion of the byte
+        for (uint8_t bit = 0; bit < 8; ++bit) {
+
+          if (!m_clock.wait(Edge::falling, 200)) {
+            return false;
+          }
+
+          // shift in from the left
+          shiftReg = shiftReg >> 1 | m_data.get() << 7;
```

Comment:
I'm not comfortable with this idea. It would work, but it is wasting a lot of ram on a small memory system just for a temporary buffer. I'd have to dig into the actual avr instructions to see how much the shifting is costing and didn't have time for that. I'm seeing good, reliable packet reading in the current code.